src/coreclr/nativeaot/Runtime/windows/PalCommon.cpp

```diff
@@ -34,6 +34,184 @@ void PalGetModuleBounds(HANDLE hOsHandle, _Out_ uint8_t ** ppLowerBound, _Out_ u
     *ppUpperBound = pbModule + cbModule - 1;
 }
 
+// Reads through the PE header of the specified module, and returns
+// the module's matching PDB's signature GUID, age, and build path by
+// fishing them out of the last IMAGE_DEBUG_DIRECTORY of type
+// IMAGE_DEBUG_TYPE_CODEVIEW.  Used when sending the ModuleLoad event
+// to help profilers find matching PDBs for loaded modules.
+//
+// Arguments:
+//
+// [in] hOsHandle - OS Handle for module from which to get PDB info
+// [out] pGuidSignature - PDB's signature GUID to be placed here
+// [out] pdwAge - PDB's age to be placed here
+// [out] wszPath - PDB's build path to be placed here
+// [in] cchPath - Number of wide characters allocated in wszPath, including NULL terminator
+//
+// This is a simplification of similar code in CLR's GetCodeViewInfo
+// in eventtrace.cpp.
+void PalGetPDBInfo(HANDLE hOsHandle, GUID * pGuidSignature, _Out_ uint32_t * pdwAge, _Out_writes_z_(cchPath) WCHAR * wszPath, int32_t cchPath)
+{
+    // Zero-init [out]-params
+    ZeroMemory(pGuidSignature, sizeof(*pGuidSignature));
+    *pdwAge = 0;
+    if (cchPath <= 0)
+        return;
+    wszPath[0] = L'\0';
+
+    BYTE *pbModule = (BYTE*)hOsHandle;
+
+    IMAGE_NT_HEADERS const * pNtHeaders = (IMAGE_NT_HEADERS*)(pbModule + ((IMAGE_DOS_HEADER*)hOsHandle)->e_lfanew);
+    IMAGE_DATA_DIRECTORY const * rgDataDirectory = NULL;
+    if (pNtHeaders->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR32_MAGIC)
+        rgDataDirectory = ((IMAGE_OPTIONAL_HEADER32 const *)&pNtHeaders->OptionalHeader)->DataDirectory;
+    else
+        rgDataDirectory = ((IMAGE_OPTIONAL_HEADER64 const *)&pNtHeaders->OptionalHeader)->DataDirectory;
+
+    IMAGE_DATA_DIRECTORY const * pDebugDataDirectory = &rgDataDirectory[IMAGE_DIRECTORY_ENTRY_DEBUG];
+
+    // In Redhawk, modules are loaded as MAPPED, so we don't have to worry about dealing
+    // with FLAT files (with padding missing), so header addresses can be used as is
+    IMAGE_DEBUG_DIRECTORY const *rgDebugEntries = (IMAGE_DEBUG_DIRECTORY const *) (pbModule + pDebugDataDirectory->VirtualAddress);
+    DWORD cbDebugEntries = pDebugDataDirectory->Size;
+    if (cbDebugEntries < sizeof(IMAGE_DEBUG_DIRECTORY))
+        return;
+
+    // Since rgDebugEntries is an array of IMAGE_DEBUG_DIRECTORYs, cbDebugEntries
+    // should be a multiple of sizeof(IMAGE_DEBUG_DIRECTORY).
+    if (cbDebugEntries % sizeof(IMAGE_DEBUG_DIRECTORY) != 0)
+        return;
+
+    // CodeView RSDS debug information -> PDB 7.00
+    struct CV_INFO_PDB70 
+    {
+        DWORD          magic; 
+        GUID           signature;       // unique identifier 
+        DWORD          age;             // an always-incrementing value 
+        _Field_z_ char  path[MAX_PATH];  // zero terminated string with the name of the PDB file 
```

Comment:
This is a stale or mangled definition of this structure. The up-to-date definition looks like this (you can check this in the MS internal OS repo) - notice that the file path is an open array:

```
struct CV_INFO_PDB70 
{
	DWORD      CvSignature; 
	GUID       Signature;       // unique identifier 
	DWORD      Age;             // an always-incrementing value 
	BYTE       PdbFileName[1];  // zero terminated string with the name of the PDB file 
};
```